Create GithubActions workflow

mumichae opened the pull request:
Including a GithubActions workflow to run the pipeline on test data

Changes:

+ changed environment files, names, R dependecy files and config files to use 'R3.6' and 'R4.0' for GA matrix
+ relies on `scib=1.0.2`, which fixes LISI parallelisation bug
+ include config files `configs/test_data-R3.6_small.yaml` and `configs/test_data-R4.0_small.yaml` for testing on subset of methods (for quicker feedback)
+ include installation script `envs/create_conda_environments.sh` for single-command installation of all environments and R methods
+ update README, moving detailed installation steps to `envs/README.md`

LuckyMD commented on envs/install_R_methods.R:
```diff
@@ -2,22 +2,36 @@
+    c("-q", "--quiet"),
+    type = 'logical',
+    default = FALSE,
+    action="store_true",
```
formatting: `n="` -> `n = "`

LuckyMD reviewed (approved):
Looks really good. A few comments to answer, otherwise seems ready to merge.

mumichae commented on envs/create_conda_environments.sh:
```diff
@@ -0,0 +1,93 @@
+    mamba env update $QUIET -f "envs/${env}.yml"
+  else
+    echo "Create $env..."
+    mamba env create $QUIET -f "envs/${env}.yml"
```
yes, so $QUIET is either '-q' or an empty string, the latter being a mamba flag

mumichae commented on .github/workflows/pipeline.yml:
```diff
@@ -0,0 +1,75 @@
+      run: |
+        mamba --version
+        ls -la
+        mamba list
```
I guess I could remove that

mumichae commented on .github/workflows/pipeline.yml:
```diff
@@ -0,0 +1,75 @@
+      shell: bash -l {0}
+      run: |
+        conda activate scib-pipeline-R${{ matrix.R }}
+        snakemake metrics --configfile configs/test_data-R${{ matrix.R }}.yaml -kc
```
From what I understand we don't need a cleanup step. Things should be reset at the end of each run.

mumichae commented on README.md:
```diff
@@ -20,153 +20,44 @@ analysis of the different steps and combinations of preprocesssing and integrati
-The main steps are:
+The installation is automated by an installation script that install the correct python and R environments based on the
+R version you want to use.
+The pipeline currently supports R 3.6 and R 4.0, and we generally recommend using the latest R version.
```
Ah, so I was trying to say that people should prefer 4.0 over 3.6 in this case. Would you still recommend R 3.6 over R 4.0? From experience updated packages tend to break older R environments, so I guess people will usually have more luck installing to newer R versions. Also, newer projects might want to use newer package versions.

LuckyMD replied:
Let's recommend 4.0 over 3.6... but not make a comment about newer versions for now. What you say might be true, it might also require us to update the code for newer versions of R

mumichae replied:
OK, I see the ambiguity. I clarified it now.

mumichae commented on README.md:
````diff
@@ -212,7 +103,7 @@ A dependency graph of the workflow can be created anytime and is useful to gain
 
 ```shell
-snakemake --configfile configs/test_data.yaml --rulegraph | dot -Tpng -Grankdir=TB > dependency.png
+snakemake --configfile configs/test_data-R3.6.yaml --rulegraph | dot -Tpng -Grankdir=TB > dependency.png
````
I think shell is only for formatting, not running ^^

mumichae commented:
Thanks @LuckyMD for the review. I addressed your comments and still need feedback on which R version to recommend.

I also added a README for the config files, since the number of files has increased in this PR. This file is not complete, so it should be revisited in a new PR, but I needed a place to describe the different config files.

Once the R question is solved, the PR should be ready to merge.

danielStrobl commented on README.md:
```diff
@@ -20,153 +20,44 @@ analysis of the different steps and combinations of preprocesssing and integrati
+To reproduce the results from this study, two separate conda environments are needed for python and R operations.
+Please make sure you have [`conda`](https://conda.io/projects/conda) installed on your system to be
+able to use the pipeline.
+We also recommend installing [`mamba`](https://mamba.readthedocs.io) for shorter installation times and smaller memory
```
Can't we just recommend mambaforge here? That's one less thing to install and has a osx-arm native version https://github.com/conda-forge/miniforge

mumichae replied:
I added mambaforge as an additional option. Many users probably already have conda, so it would be easier to just install mamba on top.

danielStrobl commented on envs/README.md:
```diff
@@ -0,0 +1,157 @@
+> changes to the pipeline code.
+> We don't recommend using it, as old dependencies tend to break after newer packages are available.
+
+## 2. Setting Environment Variables
```
Luke posted this conda package to do this automatically in the tools channel I think. Would this be an option here?

mumichae replied:
I haven't tried it yet. Might be something for a new PR though.

danielStrobl commented on envs/create_conda_environments.sh:
```diff
@@ -0,0 +1,96 @@
+#!/bin/bash
```
This looks great, I'm not sure if it's best practice to recommend people to run a shell script for installation though from a security standpoint. Maybe add something to the documentation to check scripts first before you install them?

mumichae replied:
I wasn't aware that was an actual issue. Would removing the shebang line solve this concern? I instruct users to call the script with bash anyway, so that shouldn't be a problem.

danielStrobl replied:
I meant the comment for the whole file but actually it's fine since it's not run with root privileges.

mumichae replied:
I was thinking the same. Let's keep it as is then 👍

danielStrobl reviewed (approved):
I'm not an expert on GA pipelines but as far as I can judge this looks good!
